test_phage_reference/nbconverted/1_align_quantify_samples.py

```diff
@@ -0,0 +1,211 @@
+# 
+# The fastq files store the RNA-seq results, including: sequencing and quality scores for each base call.
+# 
+# Here is a nice blog to explain how to read fastq files: https://thesequencingcenter.com/knowledge-base/fastq-files/
+# 
+# The fastq files gives the sequence of a read at a given location. Our goal is to map these reads to a reference genome so that we can quantify the number of reads that are at a given location, to determine the level of expression.
+
+# In[6]:
+
+
+if not os.path.exists(paths.FASTQ_DIR):
+    os.makedirs(paths.FASTQ_DIR)
+
+
+# In[7]:
+
+
+#!fastq-dump $paths.SRA_DIR/* --split-files --outdir $paths.FASTQ_DIR/
+
+
+# In[8]:
+
+
+# Copied from https://github.com/hoganlab-dartmouth/sraProcessingPipeline/blob/5974e040c85724a8d385e53153b7707ae7c9c255/DiscoveryScripts/quantifier.py#L83
+
+#!fastq-dump $paths_phage.SRA_DIR/* --skip-technical --readids --split-3 --clip --outdir $paths_phage.FASTQ_DIR/
+
+
+# ### Obtain a transcriptome and build an index
+# 
+# Here we are using [Salmon](https://combine-lab.github.io/salmon/)
+# 
+# **Input:**
+# * Target transcriptome
+# * This transcriptome is given to Salmon in the form of a (possibly compressed) multi-FASTA file, with each entry providing the sequence of a transcript
+# * We downloaded the phage GENOMES from NCBI GenBank
+# 
+# **Note:** For prokaryotes, transcripts and genes have a more 1-1 mapping so we're using genes for our reference transcriptome and so we don't need to use tximport to map transcript quants to genes. 
+# 
+# **Output:**
+# * The index is a structure that salmon uses to quasi-map RNA-seq reads during quantification
+# * [Quasi-map](https://academic.oup.com/bioinformatics/article/32/12/i192/2288985) is a way to map sequenced fragments (single or paired-end reads) to a target transcriptome. Quasi-mapping produces what we refer to as fragment mapping information. In particular, it provides, for each query (fragment), the reference sequences (transcripts), strand and position from which the query may have likely originated. In many cases, this mapping information is sufficient for downstream analysis like quantification.
+# 
+# *Algorithm:*
+# 
+# For a query read r through repeated application of: 
+# 1. Determining the next hash table k-mer that starts past the current query position
+# 2. Computing the maximum mappable prefix (MMP) of the query beginning with this k-mer
+# 3. Determining the next informative position (NIP) by performing a longest common prefix (LCP) query on two specifically chosen suffixes in the SA
+
+# In[9]:
+
+
+# Get PAO1 index
+get_ipython().system(' salmon index -t $paths.PAO1_REF -i $paths.PAO1_INDEX')
+
+
+# In[10]:
+
+
+# Get phage index
+get_ipython().system(' salmon index -t $paths.PHAGE_REF -i $paths.PHAGE_INDEX')
+
+
+# ### Quantify gene expression
+# Now that we have our index built and all of our data downloaded, we’re ready to quantify our samples
+# 
+# **Input:**
```

Comment:
If you have to put in multiple "input" "output" comments into a script, that violates the rule I was taught of having each script do one thing and do it well. I can understand combining them in a notebook environment, but I think building the index is different enough it should be a separate script imho.

Reply:
Yes I agree. The index only needs to be built once but the quantification can run multiple times per sample and index so I'll separate this out.